File: server/pet/transition.go
package pet

import (
	"errors"
	"math/big"
	"sort"
	"strconv"
	"strings"

	"cloud-clicker/server/fixedgrid"
)

var (
	ErrInvalidCareTransition = errors.New("invalid pet care transition")
	ErrStaleCareTransition   = errors.New("stale pet care transition")
)

type CareTransitionInput struct {
	ActionID         string `json:"action_id"`
	AttendedBeforeMS int64  `json:"attended_before_ms"`
	AttendedAfterMS  int64  `json:"attended_after_ms"`
}

type CareTransitionResult struct {
	State                  CareState           `json:"state"`
	Applied                bool                `json:"applied"`
	RejectionDetail        CareRejectionDetail `json:"rejection_detail"`
	StatID                 StatID              `json:"stat_id"`
	BeforePPM              int64               `json:"before_ppm"`
	AppliedPPM             int64               `json:"applied_ppm"`
	AfterPPM               int64               `json:"after_ppm"`
	TrustBeforePPM         int64               `json:"trust_before_ppm"`
	TrustAfterPPM          int64               `json:"trust_after_ppm"`
	Mood                   Mood                `json:"mood"`
	StatusBand             StatusBand          `json:"status_band"`
	StatusChanged          bool                `json:"status_changed"`
	NextEligibleAttendedMS int64               `json:"next_eligible_attended_ms"`
	EligibleActionIDs      []string            `json:"eligible_action_ids"`
}

func ApplyCareTransition(state CareState, catalog *Catalog, input CareTransitionInput) (CareTransitionResult, error) {
	if catalog == nil {
		return CareTransitionResult{}, ErrInvalidCareTransition
	}
	if input.AttendedBeforeMS != state.EvaluatedThroughAttendedMS || input.AttendedAfterMS < input.AttendedBeforeMS ||
		!exactNonnegative(input.AttendedAfterMS) {
		return CareTransitionResult{}, ErrStaleCareTransition
	}
	const fixturePetID = "01986666-0000-7000-8000-000000000001"
	if ValidateCareStatesForCatalog(map[string]CareState{fixturePetID: state}, catalog) != nil {
		return CareTransitionResult{}, ErrInvalidCareTransition
	}
	next := cloneCareStates(map[string]CareState{fixturePetID: state})[fixturePetID]
	statusBefore, _, err := CareStatus(next, catalog)
	if err != nil {
		return CareTransitionResult{}, err
	}
	if err := applyCareDecay(&next, catalog, input.AttendedAfterMS-input.AttendedBeforeMS); err != nil {
		return CareTransitionResult{}, err
	}
	if err := advanceBehavior(&next, catalog, input.AttendedBeforeMS, input.AttendedAfterMS); err != nil {
		return CareTransitionResult{}, err
	}
	next.EvaluatedThroughAttendedMS = input.AttendedAfterMS
	result := CareTransitionResult{State: next, RejectionDetail: RejectionUnknownAction}
	action, found := findAction(catalog, input.ActionID)
	if !found {
		return finishCareTransition(result, statusBefore, catalog, EventCareRejected)
	}
	result.StatID = action.StatID
	result.BeforePPM = next.StatsPPM[action.StatID]
	result.AfterPPM = result.BeforePPM
	result.TrustBeforePPM = next.TrustPPM
	result.TrustAfterPPM = next.TrustPPM
	result.NextEligibleAttendedMS = next.CooldownUntilAttendedMS[action.ActionID]
	switch {
	case result.NextEligibleAttendedMS > input.AttendedAfterMS:
		result.RejectionDetail = RejectionCooldown
		return finishCareTransition(result, statusBefore, catalog, EventCareRejected)
	case result.BeforePPM < action.MinEligiblePPM:
		result.RejectionDetail = RejectionIneligible
		return finishCareTransition(result, statusBefore, catalog, EventCareRejected)
	}
	effective := action.DeltaPPM
	if result.BeforePPM >= catalog.StatPolicy.DiminishingThresholdPPM {
		value := new(big.Int).Mul(big.NewInt(action.DeltaPPM), big.NewInt(catalog.StatPolicy.DiminishingFactorPPM))
		value.Quo(value, big.NewInt(1_000_000))
		effective = value.Int64()
	}
	headroom := int64(1_000_000) - result.BeforePPM
	if effective > headroom {
		result.AppliedPPM = headroom
	} else {
		result.AppliedPPM = effective
	}
	if result.AppliedPPM == 0 {
		result.RejectionDetail = RejectionSaturated
		return finishCareTransition(result, statusBefore, catalog, EventCareRejected)
	}
	if input.AttendedAfterMS > maxExactInteger-action.CooldownAttendedMS {
		return CareTransitionResult{}, ErrInvalidCareTransition
	}
	result.Applied = true
	result.RejectionDetail = ""
	result.AfterPPM = result.BeforePPM + result.AppliedPPM
	next.StatsPPM[action.StatID] = result.AfterPPM
	next.CooldownUntilAttendedMS[action.ActionID] = input.AttendedAfterMS + action.CooldownAttendedMS
	result.NextEligibleAttendedMS = next.CooldownUntilAttendedMS[action.ActionID]
	grant := catalog.TrustPolicy.GainPPMPerEffectiveAction
	if grant > catalog.TrustPolicy.CapPPM-next.TrustPPM {
		grant = catalog.TrustPolicy.CapPPM - next.TrustPPM
	}
	next.TrustPPM += grant
	result.TrustAfterPPM = next.TrustPPM
	result.State = next
	return finishCareTransition(result, statusBefore, catalog, EventCareApplied)
}

func applyCareDecay(state *CareState, catalog *Catalog, elapsedMS int64) error {
	for _, row := range catalog.StatPolicy.Stats {
		integrated, err := fixedgrid.Integrate(elapsedMS, row.DecayPPMPerGrid,
			state.StatDecayRemaindersPPM[row.StatID], catalog.StatPolicy.GridMS)
		if err != nil {
			return ErrInvalidCareTransition
		}
		headroom := state.StatsPPM[row.StatID] - row.FloorPPM
		if integrated.Whole.Cmp(big.NewInt(headroom)) >= 0 {
			state.StatsPPM[row.StatID], state.StatDecayRemaindersPPM[row.StatID] = row.FloorPPM, 0
		} else {
			state.StatsPPM[row.StatID] -= integrated.Whole.Int64()
			state.StatDecayRemaindersPPM[row.StatID] = integrated.Remainder
		}
	}
	integrated, err := fixedgrid.Integrate(elapsedMS, catalog.TrustPolicy.DecayPPMPerGrid,
		state.TrustDecayRemainderPPM, catalog.StatPolicy.GridMS)
	if err != nil {
		return ErrInvalidCareTransition
	}
	distance := state.TrustPPM - catalog.TrustPolicy.NeutralPPM
	if distance < 0 {
		distance = -distance
	}
	if integrated.Whole.Cmp(big.NewInt(distance)) >= 0 {
		state.TrustPPM, state.TrustDecayRemainderPPM = catalog.TrustPolicy.NeutralPPM, 0
	} else if state.TrustPPM > catalog.TrustPolicy.NeutralPPM {
		state.TrustPPM -= integrated.Whole.Int64()
		state.TrustDecayRemainderPPM = integrated.Remainder
	} else if state.TrustPPM < catalog.TrustPolicy.NeutralPPM {
		state.TrustPPM += integrated.Whole.Int64()
		state.TrustDecayRemainderPPM = integrated.Remainder
	} else {
		state.TrustDecayRemainderPPM = 0
	}
	return nil
}

func finishCareTransition(result CareTransitionResult, prior StatusBand, catalog *Catalog, event BehaviorEvent) (CareTransitionResult, error) {
	if err := applyBehaviorEvent(&result.State, catalog, event, result.State.EvaluatedThroughAttendedMS); err != nil {
		return CareTransitionResult{}, err
	}
	status, mood, err := CareStatus(result.State, catalog)
	if err != nil {
		return CareTransitionResult{}, err
	}
	result.StatusBand, result.Mood, result.StatusChanged = status, mood, status != prior
	result.EligibleActionIDs = EligibleCareActions(result.State, catalog, result.State.EvaluatedThroughAttendedMS)
	return result, nil
}

func CareStatus(state CareState, catalog *Catalog) (StatusBand, Mood, error) {
	if catalog == nil || len(catalog.MoodPolicy) != len(statusBands) {
		return "", "", ErrInvalidCareTransition
	}
	scalar := int64(1_000_000)
	for _, statID := range statIDs {
		if state.StatsPPM[statID] < scalar {
			scalar = state.StatsPPM[statID]
		}
	}
	selected := 0
	for index, row := range catalog.MoodPolicy {
		if row.FloorPPM <= scalar {
			selected = index
		}
	}
	return statusBands[selected], catalog.MoodPolicy[selected].MoodMember, nil
}

func EligibleCareActions(state CareState, catalog *Catalog, attendedMS int64) []string {
	if catalog == nil {
		return []string{}
	}
	result := make([]string, 0, len(catalog.Actions))
	for _, action := range catalog.Actions {
		if state.CooldownUntilAttendedMS[action.ActionID] <= attendedMS &&
			state.StatsPPM[action.StatID] >= action.MinEligiblePPM && state.StatsPPM[action.StatID] < 1_000_000 {
			result = append(result, action.ActionID)
		}
	}
	return result
}

func findAction(catalog *Catalog, actionID string) (ActionPolicy, bool) {
	index := sort.Search(len(catalog.Actions), func(index int) bool { return catalog.Actions[index].ActionID >= actionID })
	if index == len(catalog.Actions) || catalog.Actions[index].ActionID != actionID {
		return ActionPolicy{}, false
	}
	return catalog.Actions[index], true
}

func advanceBehavior(state *CareState, catalog *Catalog, before, after int64) error {
	if err := drainDue(state, before); err != nil {
		return err
	}
	nextGrid, ok := nextGridBoundary(before, catalog.StatPolicy.GridMS)
	if !ok {
		if after != before {
			return ErrInvalidCareTransition
		}
		return drainDue(state, after)
	}
	type cyclePoint struct {
		at      int64
		entered int64
	}
	seen := map[string]cyclePoint{}
	for nextGrid <= after {
		if !hasBehaviorCandidate(catalog, state.BehaviorState, EventGridTick) {
			due, exists := earliestDue(state.BehaviorQueue)
			if !exists || due > after {
				break
			}
			jump, valid := gridAtOrAfter(due, catalog.StatPolicy.GridMS)
			if !valid || jump > after {
				break
			}
			nextGrid = jump
		}
		if err := drainDue(state, nextGrid); err != nil || applyBehaviorEvent(state, catalog, EventGridTick, nextGrid) != nil {
			return ErrInvalidCareTransition
		}
		signature := behaviorCycleSignature(*state, nextGrid)
		if prior, exists := seen[signature]; exists {
			cycleMS := nextGrid - prior.at
			if cycleMS > 0 {
				cycles := (after - nextGrid) / cycleMS
				shiftEntered := state.BehaviorEnteredAtAttendedMS-prior.entered == cycleMS
				if state.BehaviorEnteredAtAttendedMS != prior.entered && !shiftEntered {
					return ErrInvalidCareTransition
				}
				for _, entry := range state.BehaviorQueue {
					if allowed := (maxExactInteger - entry.DueAttendedMS) / cycleMS; cycles > allowed {
						cycles = allowed
					}
				}
				if shiftEntered {
					if allowed := (maxExactInteger - state.BehaviorEnteredAtAttendedMS) / cycleMS; cycles > allowed {
						cycles = allowed
					}
				}
				if cycles > 0 {
					shift := cycles * cycleMS
					shiftBehaviorTimes(state, shift, shiftEntered)
					nextGrid += shift
				}
			}
		} else {
			seen[signature] = cyclePoint{at: nextGrid, entered: state.BehaviorEnteredAtAttendedMS}
		}
		if nextGrid > maxExactInteger-catalog.StatPolicy.GridMS {
			break
		}
		nextGrid += catalog.StatPolicy.GridMS
	}
	return drainDue(state, after)
}

func applyBehaviorEvent(state *CareState, catalog *Catalog, event BehaviorEvent, at int64) error {
	for _, row := range catalog.BehaviorPolicy {
		if row.FromState != state.BehaviorState || row.Event != event {
			continue
		}
		duration := new(big.Int).Mul(big.NewInt(row.DurationGridTicks), big.NewInt(catalog.StatPolicy.GridMS))
		due := new(big.Int).Add(big.NewInt(at), duration)
		if !due.IsInt64() || due.Int64() > maxExactInteger {
			return ErrInvalidCareTransition
		}
		queue := state.BehaviorQueue[:0]
		for _, entry := range state.BehaviorQueue {
			if entry.BehaviorID != string(row.ToState) {
				queue = append(queue, entry)
			}
		}
		if len(queue) >= BehaviorQueueHardcap {
			return ErrInvalidCareTransition
		}
		state.BehaviorQueue = append(queue, BehaviorQueueEntry{BehaviorID: string(row.ToState), DueAttendedMS: due.Int64()})
		sortBehaviorQueue(state.BehaviorQueue)
		return nil
	}
	return nil
}

func drainDue(state *CareState, through int64) error {
	sortBehaviorQueue(state.BehaviorQueue)
	index := 0
	for index < len(state.BehaviorQueue) && state.BehaviorQueue[index].DueAttendedMS <= through {
		entry := state.BehaviorQueue[index]
		candidate := BehaviorState(entry.BehaviorID)
		if !ValidBehaviorState(candidate) {
			return ErrInvalidCareTransition
		}
		state.BehaviorState, state.BehaviorEnteredAtAttendedMS = candidate, entry.DueAttendedMS
		index++
	}
	state.BehaviorQueue = append([]BehaviorQueueEntry(nil), state.BehaviorQueue[index:]...)
	return nil
}

func sortBehaviorQueue(queue []BehaviorQueueEntry) {
	sort.Slice(queue, func(left, right int) bool {
		if queue[left].DueAttendedMS != queue[right].DueAttendedMS {
			return queue[left].DueAttendedMS < queue[right].DueAttendedMS
		}
		return queue[left].BehaviorID < queue[right].BehaviorID
	})
}

func hasBehaviorCandidate(catalog *Catalog, state BehaviorState, event BehaviorEvent) bool {
	for _, row := range catalog.BehaviorPolicy {
		if row.FromState == state && row.Event == event {
			return true
		}
	}
	return false
}

func earliestDue(queue []BehaviorQueueEntry) (int64, bool) {
	if len(queue) == 0 {
		return 0, false
	}
	return queue[0].DueAttendedMS, true
}

func nextGridBoundary(value, grid int64) (int64, bool) {
	quotient := value / grid
	if quotient >= maxExactInteger/grid {
		return 0, false
	}
	return (quotient + 1) * grid, true
}

func gridAtOrAfter(value, grid int64) (int64, bool) {
	if value%grid == 0 {
		return value, true
	}
	if value > maxExactInteger-(grid-value%grid) {
		return 0, false
	}
	return value + grid - value%grid, true
}

func behaviorCycleSignature(state CareState, at int64) string {
	var builder strings.Builder
	builder.WriteString(string(state.BehaviorState))
	for _, entry := range state.BehaviorQueue {
		builder.WriteByte('|')
		builder.WriteString(entry.BehaviorID)
		builder.WriteByte(':')
		builder.WriteString(strconv.FormatInt(entry.DueAttendedMS-at, 10))
	}
	return builder.String()
}

func shiftBehaviorTimes(state *CareState, shift int64, shiftEntered bool) {
	if shiftEntered {
		state.BehaviorEnteredAtAttendedMS += shift
	}
	for index := range state.BehaviorQueue {
		state.BehaviorQueue[index].DueAttendedMS += shift
	}
}

File: client/src/pet/transition.ts
import { MAX_EXACT_INTEGER } from "../numeric";
import { integrateFixedGrid } from "../fixed-grid";
import type { PetCatalog } from "./catalog";
import {
  PET_BEHAVIOR_QUEUE_HARDCAP,
  PET_STAT_IDS,
  PET_STATUS_BANDS,
  type PetBehaviorEvent,
  type PetBehaviorState,
  type PetCareRejectionDetail,
  type PetMood,
  type PetStatID,
  type PetStatusBand,
} from "./grammar";
import type { PetBehaviorQueueEntry, PetCareState } from "./state";

export interface PetCareTransitionInput {
  readonly action_id: string;
  readonly attended_before_ms: number;
  readonly attended_after_ms: number;
}

export interface PetCareTransitionResult {
  readonly state: PetCareState;
  readonly applied: boolean;
  readonly rejection_detail: PetCareRejectionDetail | "";
  readonly stat_id: PetStatID | "";
  readonly before_ppm: number;
  readonly applied_ppm: number;
  readonly after_ppm: number;
  readonly trust_before_ppm: number;
  readonly trust_after_ppm: number;
  readonly mood: PetMood;
  readonly status_band: PetStatusBand;
  readonly status_changed: boolean;
  readonly next_eligible_attended_ms: number;
  readonly eligible_action_ids: readonly string[];
}

type MutableCareState = {
  stats_ppm: Record<PetStatID, number>;
  stat_decay_remainders_ppm: Record<PetStatID, number>;
  cooldown_until_attended_ms: Record<string, number>;
  trust_ppm: number;
  trust_decay_remainder_ppm: number;
  evaluated_through_attended_ms: number;
  behavior_state: PetBehaviorState;
  behavior_entered_at_attended_ms: number;
  behavior_queue: PetBehaviorQueueEntry[];
  behavior_prng_cursor: number;
};

export function applyPetCareTransition(source: PetCareState, catalog: PetCatalog, input: PetCareTransitionInput): PetCareTransitionResult {
  if (input.attended_before_ms !== source.evaluated_through_attended_ms || !safe(input.attended_after_ms) || input.attended_after_ms < input.attended_before_ms) {
    throw new RangeError("stale pet care transition");
  }
  validateCatalogState(source, catalog);
  const state = cloneState(source);
  const statusBefore = careStatus(state, catalog).status_band;
  applyDecay(state, catalog, input.attended_after_ms - input.attended_before_ms);
  advanceBehavior(state, catalog, input.attended_before_ms, input.attended_after_ms);
  state.evaluated_through_attended_ms = input.attended_after_ms;
  const action = catalog.actions.find((row) => row.action_id === input.action_id);
  const base = {
    state, applied: false, rejection_detail: "unknown_action" as PetCareRejectionDetail | "",
    stat_id: "" as PetStatID | "", before_ppm: 0, applied_ppm: 0, after_ppm: 0,
    trust_before_ppm: state.trust_ppm, trust_after_ppm: state.trust_ppm,
    next_eligible_attended_ms: 0,
  };
  if (!action) return finish(base, statusBefore, catalog, "care_rejected");
  base.stat_id = action.stat_id;
  base.before_ppm = state.stats_ppm[action.stat_id];
  base.after_ppm = base.before_ppm;
  base.next_eligible_attended_ms = state.cooldown_until_attended_ms[action.action_id] ?? 0;
  if (base.next_eligible_attended_ms > input.attended_after_ms) {
    base.rejection_detail = "cooldown";
    return finish(base, statusBefore, catalog, "care_rejected");
  }
  if (base.before_ppm < action.min_eligible_ppm) {
    base.rejection_detail = "ineligible";
    return finish(base, statusBefore, catalog, "care_rejected");
  }
  const effective = base.before_ppm >= catalog.stat_policy.diminishing_threshold_ppm
    ? Number(BigInt(action.delta_ppm) * BigInt(catalog.stat_policy.diminishing_factor_ppm) / 1_000_000n)
    : action.delta_ppm;
  base.applied_ppm = Math.min(effective, 1_000_000 - base.before_ppm);
  if (base.applied_ppm === 0) {
    base.rejection_detail = "saturated";
    return finish(base, statusBefore, catalog, "care_rejected");
  }
  if (input.attended_after_ms > MAX_EXACT_INTEGER - action.cooldown_attended_ms) throw new RangeError("pet cooldown overflow");
  base.applied = true;
  base.rejection_detail = "";
  base.after_ppm = base.before_ppm + base.applied_ppm;
  state.stats_ppm[action.stat_id] = base.after_ppm;
  state.cooldown_until_attended_ms[action.action_id] = input.attended_after_ms + action.cooldown_attended_ms;
  base.next_eligible_attended_ms = state.cooldown_until_attended_ms[action.action_id]!;
  state.trust_ppm += Math.min(catalog.trust_policy.gain_ppm_per_effective_action, catalog.trust_policy.cap_ppm - state.trust_ppm);
  base.trust_after_ppm = state.trust_ppm;
  return finish(base, statusBefore, catalog, "care_applied");
}

function applyDecay(state: MutableCareState, catalog: PetCatalog, elapsed: number): void {
  for (const row of catalog.stat_policy.stats) {
    const integrated = integrateFixedGrid(elapsed, row.decay_ppm_per_grid, state.stat_decay_remainders_ppm[row.stat_id], catalog.stat_policy.grid_ms);
    const distance = state.stats_ppm[row.stat_id] - row.floor_ppm;
    if (integrated.whole >= BigInt(distance)) {
      state.stats_ppm[row.stat_id] = row.floor_ppm;
      state.stat_decay_remainders_ppm[row.stat_id] = 0;
    } else {
      state.stats_ppm[row.stat_id] -= Number(integrated.whole);
      state.stat_decay_remainders_ppm[row.stat_id] = integrated.remainder;
    }
  }
  const integrated = integrateFixedGrid(elapsed, catalog.trust_policy.decay_ppm_per_grid, state.trust_decay_remainder_ppm, catalog.stat_policy.grid_ms);
  const distance = Math.abs(state.trust_ppm - catalog.trust_policy.neutral_ppm);
  if (integrated.whole >= BigInt(distance)) {
    state.trust_ppm = catalog.trust_policy.neutral_ppm;
    state.trust_decay_remainder_ppm = 0;
  } else if (state.trust_ppm > catalog.trust_policy.neutral_ppm) {
    state.trust_ppm -= Number(integrated.whole);
    state.trust_decay_remainder_ppm = integrated.remainder;
  } else if (state.trust_ppm < catalog.trust_policy.neutral_ppm) {
    state.trust_ppm += Number(integrated.whole);
    state.trust_decay_remainder_ppm = integrated.remainder;
  } else state.trust_decay_remainder_ppm = 0;
}

function finish(base: Omit<PetCareTransitionResult, "mood" | "status_band" | "status_changed" | "eligible_action_ids">,
  prior: PetStatusBand, catalog: PetCatalog, event: PetBehaviorEvent): PetCareTransitionResult {
  const state = base.state as MutableCareState;
  applyBehaviorEvent(state, catalog, event, state.evaluated_through_attended_ms);
  const status = careStatus(state, catalog);
  return { ...base, ...status, status_changed: status.status_band !== prior, eligible_action_ids: eligibleCareActions(state, catalog, state.evaluated_through_attended_ms) };
}

export function careStatus(state: PetCareState, catalog: PetCatalog): { readonly status_band: PetStatusBand; readonly mood: PetMood } {
  const scalar = Math.min(...PET_STAT_IDS.map((id) => state.stats_ppm[id]));
  let selected = 0;
  for (let index = 0; index < catalog.mood_policy.length; index++) if (catalog.mood_policy[index]!.floor_ppm <= scalar) selected = index;
  return { status_band: PET_STATUS_BANDS[selected]!, mood: catalog.mood_policy[selected]!.mood_member };
}

export function eligibleCareActions(state: PetCareState, catalog: PetCatalog, attendedMs: number): readonly string[] {
  return catalog.actions.filter((action) => (state.cooldown_until_attended_ms[action.action_id] ?? 0) <= attendedMs &&
    state.stats_ppm[action.stat_id] >= action.min_eligible_ppm && state.stats_ppm[action.stat_id] < 1_000_000).map((action) => action.action_id);
}

function advanceBehavior(state: MutableCareState, catalog: PetCatalog, before: number, after: number): void {
  drainDue(state, before);
  let nextGrid = nextGridBoundary(before, catalog.stat_policy.grid_ms);
  const seen = new Map<string, { readonly at: number; readonly entered: number }>();
  while (nextGrid !== null && nextGrid <= after) {
    if (!hasCandidate(catalog, state.behavior_state, "grid_tick")) {
      const due = state.behavior_queue[0]?.due_attended_ms;
      if (due === undefined || due > after) break;
      nextGrid = gridAtOrAfter(due, catalog.stat_policy.grid_ms);
      if (nextGrid === null || nextGrid > after) break;
    }
    drainDue(state, nextGrid);
    applyBehaviorEvent(state, catalog, "grid_tick", nextGrid);
    const signature = cycleSignature(state, nextGrid), prior = seen.get(signature);
    if (prior !== undefined) {
      const cycle = nextGrid - prior.at;
      if (cycle > 0) {
        let cycles = Math.floor((after - nextGrid) / cycle);
        const shiftEntered = state.behavior_entered_at_attended_ms - prior.entered === cycle;
        if (state.behavior_entered_at_attended_ms !== prior.entered && !shiftEntered) throw new RangeError("noncanonical pet behavior cycle");
        for (const entry of state.behavior_queue) cycles = Math.min(cycles, Math.floor((MAX_EXACT_INTEGER - entry.due_attended_ms) / cycle));
        if (shiftEntered) cycles = Math.min(cycles, Math.floor((MAX_EXACT_INTEGER - state.behavior_entered_at_attended_ms) / cycle));
        if (cycles > 0) {
          const shift = cycles * cycle;
          if (shiftEntered) state.behavior_entered_at_attended_ms += shift;
          state.behavior_queue = state.behavior_queue.map((entry) => ({ ...entry, due_attended_ms: entry.due_attended_ms + shift }));
          nextGrid += shift;
        }
      }
    } else seen.set(signature, { at: nextGrid, entered: state.behavior_entered_at_attended_ms });
    nextGrid = nextGrid > MAX_EXACT_INTEGER - catalog.stat_policy.grid_ms ? null : nextGrid + catalog.stat_policy.grid_ms;
  }
  drainDue(state, after);
}

function applyBehaviorEvent(state: MutableCareState, catalog: PetCatalog, event: PetBehaviorEvent, at: number): void {
  const row = catalog.behavior_policy.find((candidate) => candidate.from_state === state.behavior_state && candidate.event === event);
  if (!row) return;
  const dueWide = BigInt(at) + BigInt(row.duration_grid_ticks) * BigInt(catalog.stat_policy.grid_ms);
  if (dueWide > BigInt(MAX_EXACT_INTEGER)) throw new RangeError("pet behavior due overflow");
  const queue = state.behavior_queue.filter((entry) => entry.behavior_id !== row.to_state);
  if (queue.length >= PET_BEHAVIOR_QUEUE_HARDCAP) throw new RangeError("pet behavior queue full");
  queue.push({ behavior_id: row.to_state, due_attended_ms: Number(dueWide) });
  queue.sort(compareQueue);
  state.behavior_queue = queue;
}

function drainDue(state: MutableCareState, through: number): void {
  state.behavior_queue.sort(compareQueue);
  let index = 0;
  while (index < state.behavior_queue.length && state.behavior_queue[index]!.due_attended_ms <= through) {
    const entry = state.behavior_queue[index]!;
    state.behavior_state = entry.behavior_id as PetBehaviorState;
    state.behavior_entered_at_attended_ms = entry.due_attended_ms;
    index++;
  }
  state.behavior_queue = state.behavior_queue.slice(index);
}

function compareQueue(left: PetBehaviorQueueEntry, right: PetBehaviorQueueEntry): number {
  return left.due_attended_ms - right.due_attended_ms || byteCompare(left.behavior_id, right.behavior_id);
}

function hasCandidate(catalog: PetCatalog, state: PetBehaviorState, event: PetBehaviorEvent): boolean {
  return catalog.behavior_policy.some((row) => row.from_state === state && row.event === event);
}

function nextGridBoundary(value: number, grid: number): number | null {
  const result = (Math.floor(value / grid) + 1) * grid;
  return Number.isSafeInteger(result) && result <= MAX_EXACT_INTEGER ? result : null;
}

function gridAtOrAfter(value: number, grid: number): number | null {
  const result = Math.ceil(value / grid) * grid;
  return Number.isSafeInteger(result) && result <= MAX_EXACT_INTEGER ? result : null;
}

function cycleSignature(state: MutableCareState, at: number): string {
  return `${state.behavior_state}${state.behavior_queue.map((entry) => `|${entry.behavior_id}:${entry.due_attended_ms - at}`).join("")}`;
}

function cloneState(state: PetCareState): MutableCareState {
  return { ...state, stats_ppm: { ...state.stats_ppm }, stat_decay_remainders_ppm: { ...state.stat_decay_remainders_ppm },
    cooldown_until_attended_ms: { ...state.cooldown_until_attended_ms }, behavior_queue: state.behavior_queue.map((entry) => ({ ...entry })) };
}

function validateCatalogState(state: PetCareState, catalog: PetCatalog): void {
  if (state.behavior_prng_cursor !== 0 || state.trust_decay_remainder_ppm >= catalog.stat_policy.grid_ms ||
    PET_STAT_IDS.some((id) => state.stat_decay_remainders_ppm[id] >= catalog.stat_policy.grid_ms)) throw new RangeError("invalid pet state for catalog");
}

function safe(value: number): boolean {
  return Number.isSafeInteger(value) && value >= 0 && value <= MAX_EXACT_INTEGER;
}

function byteCompare(left: string, right: string): number {
  const a = new TextEncoder().encode(left), b = new TextEncoder().encode(right);
  for (let index = 0; index < Math.min(a.length, b.length); index++) if (a[index] !== b[index]) return a[index]! - b[index]!;
  return a.length - b.length;
}
